clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java

```diff
@@ -207,6 +226,232 @@ public NetworkClientDelegate.UnsentRequest toUnsentRequest() {
+         *  <li>1. dedupe against unsents: if a duplicated request was previously made, we chain the future</>
+         *  <li>2. dedupe against incompleted: If a duplicated request was sent but hasn't gotten a results, we chain
+         *  the future.</>
+         *
+         *  <p>If the request is new, we chain a call back to remove itself from the {@code inflightOffsetFetches}
+         *  upon completion.</>
+         */
+        private CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> addOffsetFetchRequest(final UnsentOffsetFetchRequest request) {
+            Optional<UnsentOffsetFetchRequest> dupe =
+                    unsentOffsetFetches.stream().filter(r -> r.sameRequest(request)).findAny();
+            Optional<UnsentOffsetFetchRequest> inflight =
+                    inflightOffsetFetches.stream().filter(r -> r.sameRequest(request)).findAny();
+
+            if (dupe.isPresent() || inflight.isPresent()) {
+                log.info("Duplicated OffsetFetchRequest: " + request.requestedPartitions);
+                dupe.orElseGet(() -> inflight.get()).chainFuture(request.future);
+            } else {
+                // remove the request from the outbound buffer: inflightOffsetFetches
+                request.future.whenComplete((r, t) -> {
+                    if (!inflightOffsetFetches.remove(request)) {
+                        log.info("unable to remove request from the outbound buffer:" + request);
+                    }
+                });
+                this.unsentOffsetFetches.add(request);
+            }
+            return request.future;
+        }
+
+        private CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> addOffsetFetchRequest(final Set<TopicPartition> partitions) {
+            UnsentOffsetFetchRequest request = new UnsentOffsetFetchRequest(
+                    partitions,
+                    groupState.generation,
+                    retryBackoffMs);
+            return addOffsetFetchRequest(request);
+        }
+
+        public List<NetworkClientDelegate.UnsentRequest> drain(final long currentTimeMs) {
+            List<NetworkClientDelegate.UnsentRequest> unsent = new ArrayList<>();
+            unsent.addAll(unsentOffsetCommits.stream()
+                    .map(UnsentOffsetCommitRequest::toUnsentRequest)
+                    .collect(Collectors.toList()));
+            List<UnsentOffsetFetchRequest> sendables = unsentOffsetFetches.stream()
+                    .filter(r -> r.canSendRequest(currentTimeMs))
+                    .collect(Collectors.toList());
+            inflightOffsetFetches.addAll(sendables);
+            unsent.addAll(sendables.stream()
+                    .peek(r -> r.onSendAttempt(currentTimeMs))
+                    .map(r -> r.toUnsentRequest(currentTimeMs))
+                    .collect(Collectors.toList()));
+            // empty the staged requests
+            unsentOffsetCommits.clear();
```

Comment:
This seems not right to me as well: if an unsent request is filtered in `.filter(r -> r.canSendRequest(currentTimeMs))` they should still be kept in the queue, right?

Reply:
So `unsentOffsetCommits` and `unsentOffsetFetches` tracks all of the requests that haven't been returned by the poll call (which subsequently make to the NetworkClient).  So the poll calls drain() in the end, and dumping the unsentRequest into the `inflightOffsetFetches`.

I think the naming and the way the code is structured is probably too confusing. Maybe there's a better way to bookkeep these inflight/unsent requests.

Reply:
seems like a bug here: I cleared the non-sendables.

Reply:
Yeah, I think we should not just blindly empty the `unsentOffsetFetches`, instead we should only remove `sendables` from the queue. Could you fix it?

Reply:
Ah yup! I was not watching the latest commits you had.